pallets/loans/src/interest.rs

```diff
@@ -32,52 +32,105 @@ impl<T: Config> Pallet<T> {
         if now <= last_accrued_interest_time {
             return Ok(());
         }
+
+        let (
+            borrow_rate,
+            supply_rate,
+            exchange_rate,
+            util,
+            total_borrows_new,
+            total_reserves_new,
+            borrow_index_new,
+        ) = Self::get_market_status(asset_id)?;
+
         Self::update_last_accrued_interest_time(asset_id, now)?;
-        let delta_time = now - last_accrued_interest_time;
+        TotalBorrows::<T>::insert(asset_id, total_borrows_new);
+        TotalReserves::<T>::insert(asset_id, total_reserves_new);
+        BorrowIndex::<T>::insert(asset_id, borrow_index_new);
+
+        //save redundant storage right now.
+        UtilizationRatio::<T>::insert(asset_id, util);
+        BorrowRate::<T>::insert(asset_id, borrow_rate);
+        SupplyRate::<T>::insert(asset_id, supply_rate);
+        ExchangeRate::<T>::insert(asset_id, exchange_rate);
+
+        Ok(())
+    }
+
+    pub fn get_market_status(
+        asset_id: AssetIdOf<T>,
+    ) -> Result<
+        (
+            Rate,
+            Rate,
+            Rate,
+            Ratio,
+            BalanceOf<T>,
+            BalanceOf<T>,
+            FixedU128,
+        ),
+        DispatchError,
+    > {
         let market = Self::market(asset_id)?;
+        let total_supply = Self::total_supply(asset_id);
         let total_cash = Self::get_total_cash(asset_id);
-        let total_borrows = Self::total_borrows(asset_id);
-        let total_reserves = Self::total_reserves(asset_id);
+        let mut total_borrows = Self::total_borrows(asset_id);
+        let mut total_reserves = Self::total_reserves(asset_id);
+        let mut borrow_index = Self::borrow_index(asset_id);
+
         let util = Self::calc_utilization_ratio(total_cash, total_borrows, total_reserves)?;
         let borrow_rate = market
             .rate_model
             .get_borrow_rate(util)
             .ok_or(ArithmeticError::Overflow)?;
         let supply_rate =
             InterestRateModel::get_supply_rate(borrow_rate, util, market.reserve_factor);
-        UtilizationRatio::<T>::insert(asset_id, util);
-        BorrowRate::<T>::insert(asset_id, &borrow_rate);
-        SupplyRate::<T>::insert(asset_id, supply_rate);
-        Self::update_borrow_index(borrow_rate, asset_id, &market, delta_time)?;
-        Self::update_exchange_rate(asset_id)?;
 
-        Ok(())
+        let now = T::UnixTime::now().as_secs();
+        let last_accrued_interest_time = Self::last_accrued_interest_time(asset_id);
+        if now > last_accrued_interest_time {
```

Comment:
I think we should return Err once `now <= last_accrued_interest_time` rather than continue calculate the exchange_rate.

Reply:
get_market_status is just read_only function. And if we return Err, that means if accruedInterest happened in Block 1000, and we can not fetch the status at Block 1000. That's not reasonable.

Reply:
OK, got it.